server/aesdsocket.c

```diff
@@ -95,48 +125,68 @@ void *get_in_addr(struct sockaddr *sa)
 }
 
 /**
- * @brief Perform global clean-up operations.
+ * @brief Perform global clean-up operations for a multi-threaded application.
  * 
- * This function handles the clean-up tasks for the application, including closing
- * files, deleting temporary data, and closing socket and client descriptors.
- * It also closes the syslog.
+ * This function takes care of closing files, sockets, and client descriptors, 
+ * deleting temporary data, freeing memory, and performing other clean-up tasks.
+ * It also handles logging for these operations.
  */
 void cleanup_on_exit(void)
 {
-	int ret_status;
-	
-	// Close data file
-	ret_status = close(dataFileDescriptor);
-	if(ret_status == ERROR)
-	{
-		syslog(LOG_ERR,"File close failed");
-		s_flags.command_status_success = false;
-	}
-	
-	// delete data file
-	ret_status = unlink(DATA_FILE);
-	if(ret_status == ERROR)
-	{
-		syslog(LOG_ERR,"File delete failed");
-	}
-	
+    int ret_status;
+
+    // Log initiation of cleanup
+    syslog(LOG_INFO, "Initiating clean-up procedures.");
+
+    // Close data file
+    ret_status = close(dataFileDescriptor);
+    if(ret_status == -1)
+    {
+        syslog(LOG_ERR, "Failed to close data file.");
+        s_flags.command_status_success = false;
+    }
+
+    // Delete data file
+    ret_status = unlink(DATA_FILE);
+    if(ret_status == -1)
+    {
+        syslog(LOG_ERR, "Failed to delete data file.");
+    }
+
+    // Free elements from the queue if any (assuming head is defined elsewhere)
+    while (!SLIST_EMPTY(&head))
+    {
+        node = SLIST_FIRST(&head);
+        SLIST_REMOVE(&head, node, node, nodes);
+        free(node);
+        node = NULL;
+    }
+
+    // Join timestamp thread
+    pthread_join(TS_data.threadId, NULL);
+
+    // Destroy mutex lock (assuming lock is defined elsewhere)
+    pthread_mutex_destroy(&lock);
```

Comment:
@Suraj-Ajjampur is this safe to do in a signal handler?

Reply:
I understand that these functions are not asynchronous-signal-safe according to POSIX standards. Using non-async-signal-safe functions inside a signal handler can lead to undefined behavior, such as deadlocks or corruption.

With the latest changes, I am enabling a flag in the handle_termination function before calling the cleanup_on_exit() function.

In the cleanup function, I am checking for that flag, before performing these operations

```
    if(s_flags.signal_caught != true){
    // Join timestamp thread
    pthread_join(TS_data.threadId, NULL);

    // Destroy mutex lock (assuming lock is defined elsewhere)
    pthread_mutex_destroy(&lock);
    }
```